microsoft/onefuzz
# Scoped notification pause

## Summary of the Pull Request

_What is this about?_

* Adds capability to pause notification creation at the container level
* Closes #3556 

## Info on Pull Request

_What does this include?_

* New containers endpoint for modifying container metadata
* New CLI command for modifying container metadata (`onefuzz containers update {container name} {metadata key}={metadata value}`)
* Fixes a bug where messages requeued due to `ADO_WORKITEM_PROCESSING_DISABLED` have their retry count incremented. We should only increment the retry count when we actually _try_ to process the notification
* Do not attempt to process notifications for containers that are tagged with `pauseNotifications=true`. If that's the case, requeue the message with a 1 day timeout

## Validation Steps Performed

_How does someone test & validate?_

Tested in dev deployment:

* Create a file in a container
* Check that a message was added in file-changes queue
* Using the cli, set pauseNotifications=true metadata tag on the container
* Create another file in the container
* Check that a message was added in file-changes queue
* Check that file-changes queue shows "0 of 1 messages" meaning that the message was requeued with a visibility timeout
